status/status.go

```diff
@@ -128,3 +130,39 @@ func FromError(err error) (s *Status, ok bool) {
 	}
 	return nil, false
 }
+
+// WithDetails returns a new status with the provided details messages appended to the status.
+// If any errors are encountered, it returns nil and the first error encountered.
+func (s *Status) WithDetails(details ...proto.Message) (*Status, error) {
+	if s.Code() == codes.OK {
+		return nil, errors.New("no error details for status with code OK")
+	}
+	// s.Code() != OK implies that s.Proto() != nil.
+	p := s.Proto()
+	for _, detail := range details {
+		any, err := ptypes.MarshalAny(detail)
+		if err != nil {
+			return nil, err
+		}
+		p.Details = append(p.Details, any)
+	}
+	return &Status{s: p}, nil
+}
+
+// Details returns a slice of details messages attached to the status.
+// If any errors are encountered, it returns nil and the first error encountered.
+func (s *Status) Details() ([]proto.Message, error) {
+	if s.Code() == codes.OK {
+		return nil, errors.New("no error details for status with code OK")
+	}
+	// s.Code() != OK implies that s.s != nil.
+	details := make([]proto.Message, 0, len(s.s.Details))
+	for _, any := range s.s.Details {
+		detail := &ptypes.DynamicAny{}
+		if err := ptypes.UnmarshalAny(any, detail); err != nil {
+			return nil, err
+		}
```

Comment:
My comment was on the `WithDetails` method.  I can't think of any good reason to have that return a not-fully-populated status.

I agree it is possible that `Details` may have useful partial results.  But if we do that, I wouldn't want to stop when the first error is encountered -- the order would be important when it shouldn't be.  Instead I'd want to return all of the messages and all of the errors encountered.  A few ideas:

1. `func (s *Status) Details() []struct{proto.Message, error}`

I.e. return a slice of a struct containing _either_ the message or the error encountered while decoding it.

2. `func (s *Status) Details() []proto.Message`

Return `nil` for entries in the details that encountered an error decoding (or omit them).

3. `func (s *Status) Details() []interface{}`

The values returned in the slice would be documented as either a proto.Message or an error. 
 Type assertion is going to be used already by 99% of users of `Details`, so maybe this is pretty reasonable.